### PR TITLE
improve optimizer rule "remove-filters-covered-by-index"

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Make optimizer rule "remove-filters-covered-by-index" remove FILTERs that were referring to aliases of the collection variable, e.g.
```
FOR doc IN collection
  LET value = doc.indexedAttribute
  FILTER value == ...
```
Previously, FILTERs that were using aliases were not removed by that optimizer rule.
In addition, the optimizer rule "remove-unnecessary-calculations" will now run again in case it successfully removed variables. This can unlock further removal of unused variables in sequences such as
```
FOR doc IN collection
  LET value = doc.indexedAttribute
  LET tmp1 = value > ...
  LET tmp2 = value < ...
```
when the removal of `tmp1` and `tmp2` makes it possible to also remove the calculation of `value`.

- [ ] :hankey: Bugfix 
- [x] :pizza: New feature 
- [ ] :hammer: Refactoring 
- [x] :book: CHANGELOG entry made
- [x] :muscle: The behavior in this PR was *manually tested*
- [x] :computer: The behavior change can be verified via automatic tests

#### Backports:

- [x] No backports required
- [ ] Backports required for: *(Please specify versions)*

### Testing & Verification

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [x] This change is already covered by existing tests, such as *AQL tests*.
- [x] This PR adds tests that were used to verify all changes:
  - [ ] Added **Regression Tests**
  - [ ] Added new C++ **Unit Tests**
  - [x] Added new **integration tests** (i.e. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)

http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11493/

### CODE REVIEW

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11511/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11517/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11577/

File: arangod/Aql/OptimizerRules.cpp [hunk 1]
```diff
@@ -2911,9 +2893,17 @@ void arangodb::aql::removeUnnecessaryCalculationsRule(Optimizer* opt,
 
   if (!toUnlink.empty()) {
     plan->unlinkNodes(toUnlink);
+    if (nodes.size() > 1) {
```
Comment (graetzer) [inline arangod/Aql/OptimizerRules.cpp:2896]: isn't this only true if you have more remaining CalculationNodes `(nodes.size() - toUnlink.size()) > 0`

**Review by graetzer — APPROVED** [review]

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/11624/

File: arangod/Aql/OptimizerRules.cpp (hunk 1 above)
Comment (jsteemann) [reply arangod/Aql/OptimizerRules.cpp]: Looks like. I made this change, and also found a bug in the if...then...else here.
Re-running it in Jenkins now...

**Comment by jsteemann** [comment]
Tests blue again

**Review by graetzer — APPROVED** [review]

**Pull request merged**